### PR TITLE
leap: More test cases for years in the past (less than four digits)

### PR DESCRIPTION
**Pull request by jonasbn**
Having observed basic solutions on other Rust exercises ignoring the fact that years might not necessarily only be four digit, bys adding tests for years in the past might assist the aspiring Rust programmer.

### CODE REVIEW

**Comment by coriolinus** [comment]
This is a good idea, but my inclination is to not review this while https://github.com/exercism/problem-specifications/pull/1581 is pending. Once that's merged, it would be good to add/adjust tests as required to bring us into compliance with the spec, instead of inventing our own tests.

**Comment by jonasbn** [comment]
I was thinking about interesting corner cases and the year `0` comes to mind.

The following test case, could be considered for addition.

```rust
assert_eq!(leap::is_leap_year(0), false);
```

At the same time, it does make the solution more complex, but also demonstrates good test and coding practice.

The value `0` is completely legit, but does make an expected reference implementation fail/explode if you are a _n00b_ like me.

**Comment by kotp** [comment]
> The value 0 is completely legit, but does make an expected reference implementation fail/explode if you are a n00b like me.

Is the year 0 legitimate?  I thought it did not exist in the Gregorian calendar.  I know it does in other calendars.  The Leap Year rules seem to follow Gregorian rules, rather than the simpler Julian rules.

File: exercises/leap/tests/leap.rs [hunk 1]
```diff
@@ -37,6 +37,11 @@ fn test_year_divisible_by_400_leap_year() {
 #[ignore]
 fn test_any_old_year() {
     assert_eq!(leap::is_leap_year(1997), false);
+    assert_eq!(leap::is_leap_year(1), false);
+    assert_eq!(leap::is_leap_year(4), true);
```
[the remark this answers is not in the document]
Comment (kotp) [reply exercises/leap/tests/leap.rs]: > The algorithm is not necessarily valid for years in the Julian calendar, such as years before 1752 in the British Empire. The year 1700 was a leap year in the Julian calendar, but not in the Gregorian calendar.

In the spirit of the exercise, given Gregorian calendar specification in the Readme, perhaps introducing a specification that prior to 1582 it should fail.  As in ordinary use, we don't go back before that date, and when going back, we have to make adjustments.  Let's keep the exercise simple, it is about logic, more than dates.  Calendaring opens up a lot of nuances, and I believe the Gregorian calendar was used with this in mind, and as a limitation of expectation.

And since the exercise is not about date display, I don't think it makes sense to apply ISO 8601 since that is a display standard.  We are reporting true or false, only, right?

> The reason why I raised the issue is not due to the Gregorian calendar mention, but more the programming issues with the assignment. The assignment uses an integer for implementation which allows 0 and values to the maximum available size, so a programmer would have to handle the values and corner cases no matter the interpretation of the Gregorian calendar.

Then let's limit the integer use, maybe.  A test that holds the upper bound the Gregorian calendar is meant to be used for.  And a lower bound that starts at adoption.

Otherwise we really do need to introduce everything that has been hinted at in this discussion, and even more comes to mind, on a global level, how do we get the corresponding date for the other common calendars of the world?  Do they even have the idea of a leap year?  Since this exercise isn't "find out the date", but "is it a leap year?" I think we can restrict the use inside the Gregorian calandar, and since the idea is to not use the built in Date facilities, we don't need to worry about the different display formats.

On the other hand, the subject is extremely interesting, and I might enjoy exploring date conversions, finding out of a date was a leap year in year -4.  But it should have been actually a leap year in history, not what would have been a leap year based on Gregorian rules.

[the remark this answers is not in the document]
Comment (jonasbn) [reply exercises/leap/tests/leap.rs]: I have put the tests into a separate tests. I have left out the test of year zero.

**Comment by kotp** [comment]
> This looks good to me. Thanks, @jonasbn!
Make no mistake, I appreciate the work as well @jonasbn even if I have reservations about introducing the early years, pre-adoption.  And especially since it will effect, potentially, all the tracks that use this exercise.

> @kotp, you do have veto power if you think we really shouldn't add tests for early years. It sounds like your objections are related to the philosophy of dates rather than the actual implementation.
 
Yes, the objections are strictly on the subject of Dates, it is not the goal of the exercise to adequately address years prior to 1572, nor address regional adoption times.  The ability to talk about this, especially if a student mentions that some year is a leap year, we can surely advise them about the core or standard libraries available (whatever the term is for the language) that there is a lot of work done to help us ensure that we get things mostly right, by using those tools.  This exercise is more about how to manage the logic statement, than actually compute "leap year" to any degree of reliability.

If we add information that relates that the tests do not drive a canonical definition, and that we can't reasonably answer "bar bets" with this tool, or we don't mention it and also don't test early years.

The exercise is eye opening enough when there is a curious student that tends to ask questions about centuries, such as 1500 and beyond... I generally refer them to check answers against the implemented tooling in the core and standard libraries, to see how the naive leap year solution measures up.  And as an educational outlet, we should be careful to not "outright lie" as someone may find themselves arguing one day "But I learned that year XXX was a leap year, because the tests from Exercism had it like that."  

An interesting read about Julian calendar dates, and Gregorian calendar dates [is here](https://tlehrian.wordpress.com/2008/02/29/totally-useless-leap-year-facts/).  It is usually where I start people reading when they express an interest.  At least for the last decade or so.

**Comment by kotp** [comment]
We might wonder if it would be interesting to "toggle the rules to Julian" if the lower bound does hit 1751 and lower... that may be an interesting aspect, and would provide a pointer to information regarding date.  That is about where I would want to leave the tests.

**Comment by coriolinus** [comment]
@kotp I am interpreting your statements as "I have reservations, but not enough to veto this PR." 

I agree that as educators we should avoid falsehoods. However, I believe simplifications are permissible. Example; we have several exercises talking about crypto, and no student should ever use any implementation of those to do actual cryptography with. To me, sticking with the simple definition of a leap year is an allowable simplification. Testing a wide range of years is expanding coverage of our simplification, not arguing that the simplification actually works in all times and places.

**Comment by kotp** [comment]
OK, but that isn't what I said.

And given your example, we are missing the "warnings" that are similar to the warnings we should be giving people on the "crypto" exercises as well.  Don't use this for years prior to Gregorian Calendar years, and don't use Crypto that you devise for doing crypto.

The simple definition of leap years was the Julian calendar, not the "slightly less simple" definition of what this is.

But no, I wouldn't veto, the tests give us talking points for the students that want to explore it a little deeper.  Otherwise we should take away the mentioning of the Gregorian calendar, since our rules now don't have that lower bound.

No, I wouldn't veto the change, but now, given your example, we should give fair warning, just as we would for a cryptographical exercise.
